File: Sources/HarmonyTrainer/ChordDetection.swift
import Foundation

public enum ChordDetection {
    /// Detect the best-matching chord from a set of MIDI note numbers.
    /// Returns nil if fewer than 3 distinct pitch classes or no match found.
    public static func detect(from midiNotes: [Int]) -> Chord? {
        let pitchClassValues = Set(midiNotes.map { (($0 % 12) + 12) % 12 })
        guard pitchClassValues.count >= 3 else { return nil }

        let sortedPC = pitchClassValues.sorted()
        let bassNote = midiNotes.min() ?? 0
        let bassPitchClass = ((bassNote % 12) + 12) % 12

        for chordType in ChordType.allTypes {
            for potentialRoot in sortedPC {
                let expectedIntervals = Set(chordType.intervals)
                let actualIntervals = Set(sortedPC.map { ($0 - potentialRoot + 12) % 12 })

                guard actualIntervals == expectedIntervals else { continue }

                let root = PitchClass(rawValue: potentialRoot)!
                let inversion = determineInversion(
                    root: potentialRoot,
                    bassPC: bassPitchClass,
                    intervals: chordType.intervals
                )
                return Chord(root: root, chordType: chordType, inversion: inversion)
            }
        }

        return nil
    }

    /// Returns all possible chord interpretations ranked by likelihood.
    public static func detectAll(from midiNotes: [Int]) -> [Chord] {
        let pitchClassValues = Set(midiNotes.map { (($0 % 12) + 12) % 12 })
        guard pitchClassValues.count >= 3 else { return [] }

        let sortedPC = pitchClassValues.sorted()
        let bassNote = midiNotes.min() ?? 0
        let bassPitchClass = ((bassNote % 12) + 12) % 12
        var results: [Chord] = []

        for chordType in ChordType.allTypes {
            for potentialRoot in sortedPC {
                let expectedIntervals = Set(chordType.intervals)
                let actualIntervals = Set(sortedPC.map { ($0 - potentialRoot + 12) % 12 })

                guard actualIntervals == expectedIntervals else { continue }

                let root = PitchClass(rawValue: potentialRoot)!
                let inversion = determineInversion(
                    root: potentialRoot,
                    bassPC: bassPitchClass,
                    intervals: chordType.intervals
                )
                results.append(Chord(root: root, chordType: chordType, inversion: inversion))
            }
        }

        return results
    }

    private static func determineInversion(root: Int, bassPC: Int, intervals: [Int]) -> Inversion {
        guard bassPC != root else { return .root }
        let chordPCs = intervals.map { (root + $0) % 12 }
        guard let bassIndex = chordPCs.firstIndex(of: bassPC) else { return .root }
        switch bassIndex {
        case 1: return .first
        case 2: return .second
        case 3: return .third
        default: return .root
        }
    }
}

File: web/src/domain/chordDetection.ts
import { type PitchClass } from './pitchClass.js';
import { ALL_CHORD_TYPES } from './chordType.js';
import { Inversion } from './inversion.js';
import { type Chord, createChord } from './chord.js';

function determineInversion(root: number, bassPC: number, intervals: readonly number[]): Inversion {
  if (bassPC === root) return Inversion.Root;
  const chordPCs = intervals.map(i => (root + i) % 12);
  const bassIndex = chordPCs.indexOf(bassPC);
  switch (bassIndex) {
    case 1: return Inversion.First;
    case 2: return Inversion.Second;
    case 3: return Inversion.Third;
    default: return Inversion.Root;
  }
}

export function detect(midiNotes: number[]): Chord | null {
  const pitchClassValues = new Set(midiNotes.map(n => (((n % 12) + 12) % 12)));
  if (pitchClassValues.size < 3) return null;

  const sortedPC = [...pitchClassValues].sort((a, b) => a - b);
  const bassNote = Math.min(...midiNotes);
  const bassPitchClass = ((bassNote % 12) + 12) % 12;

  for (const chordType of ALL_CHORD_TYPES) {
    for (const potentialRoot of sortedPC) {
      const expectedIntervals = new Set(chordType.intervals);
      const actualIntervals = new Set(sortedPC.map(pc => ((pc - potentialRoot) + 12) % 12));

      if (expectedIntervals.size !== actualIntervals.size) continue;
      let match = true;
      for (const v of expectedIntervals) {
        if (!actualIntervals.has(v)) { match = false; break; }
      }
      if (!match) continue;

      const root = potentialRoot as PitchClass;
      const inversion = determineInversion(potentialRoot, bassPitchClass, chordType.intervals);
      return createChord(root, chordType, inversion);
    }
  }

  return null;
}

export function detectAll(midiNotes: number[]): Chord[] {
  const pitchClassValues = new Set(midiNotes.map(n => (((n % 12) + 12) % 12)));
  if (pitchClassValues.size < 3) return [];

  const sortedPC = [...pitchClassValues].sort((a, b) => a - b);
  const bassNote = Math.min(...midiNotes);
  const bassPitchClass = ((bassNote % 12) + 12) % 12;
  const results: Chord[] = [];

  for (const chordType of ALL_CHORD_TYPES) {
    for (const potentialRoot of sortedPC) {
      const expectedIntervals = new Set(chordType.intervals);
      const actualIntervals = new Set(sortedPC.map(pc => ((pc - potentialRoot) + 12) % 12));

      if (expectedIntervals.size !== actualIntervals.size) continue;
      let match = true;
      for (const v of expectedIntervals) {
        if (!actualIntervals.has(v)) { match = false; break; }
      }
      if (!match) continue;

      const root = potentialRoot as PitchClass;
      const inversion = determineInversion(potentialRoot, bassPitchClass, chordType.intervals);
      results.push(createChord(root, chordType, inversion));
    }
  }

  return results;
}
